File: 3215-matrix-similarity-after-cyclic-shifts/matrix-similarity-after-cyclic-shifts.py
class Solution:

    def areSimilar(self, mat: List[List[int]], k: int) -> bool:
        def left_shift(row):
            return row[1:] + [row[0]]
        def right_shift(row):
            return [row[-1]] + row[:-1]
        original_matrix = [row[:] for row in mat]

        for _ in range(k):
            for i in range(len(mat)):
                if i % 2 == 0:
                    mat[i] = left_shift(mat[i])
                else:
                    mat[i] = right_shift(mat[i])
        return mat == original_matrix
        

File: 3215-matrix-similarity-after-cyclic-shifts/matrix-similarity-after-cyclic-shifts.js
/**
 * @param {number[][]} mat
 * @param {number} k
 * @return {boolean}
 */
var areSimilar = function(mat, k) {
    const n = mat[0].length;
    const shift = k % n;
    if (shift === 0) {
        return true;
    }

    for (let i = 0; i < mat.length; i++) {
        for (let j = 0; j < n; j++) {
            let shiftIndex;
            if (i % 2 === 0) {
                shiftIndex = (j + shift) % n;
            } else {
                shiftIndex = (j - shift + n) % n;
            }

            if (mat[i][j] !== mat[i][shiftIndex]) {
                return false
            }
        }
    }
    return true;
};